Source/WebInspectorUI/UserInterface/Views/FontVariationDetailsSectionRow.js

```diff
@@ -142,6 +165,87 @@ WI.FontVariationDetailsSectionRow = class FontVariationDetailsSectionRow extends
 
         return 1;
     }
+
+    _showValidity()
+    {
+        if (!this._hasValidValue)
+            this.warningMessage = WI.UIString("Axis value outside of supported range: %s – %s", "A warning that is shown in the Font Details Sidebar when the value for a variation axis is outside of the supported range of values").format(this._formatAxisValueAsString(this._minimumValue), this._formatAxisValueAsString(this._maximumValue));
+        else
+            this.warningMessage = null;
+    }
+
+    _handleValueBlurEvent(event)
+    {
+        this._skipUpdatingInputValue = false;
+        this.value = this._value;
+    }
+
+    _handleValueInputEvent(event)
+    {
+        if (this._showValidityTimeout)
+            clearTimeout(this._showValidityTimeout);
+
+        // Allow the input to temporarily hold an invalid value without immediately warning the user.
+        this._showValidityTimeout = setTimeout(this._showValidity.bind(this), 750);
+
+        let valueAsString = event.target.value;
+        let valueAsNumber = event.target.valueAsNumber;
+
+        this._hasValidValue = false;
+        const validValueFormat = /\-?\d+(.\d+)?/g;
+
+        if (!validValueFormat.test(valueAsString)) {
+            this._skipUpdatingInputValue = true;
+
+            this.dispatchEventToListeners(WI.FontVariationDetailsSectionRow.Event.VariationValueChanged, {tag: this._tag, value: this._defaultValue});
+            return;
+        }
+
+        if (!(this._minimumValue <= valueAsNumber && valueAsNumber <= this._maximumValue)) {
+            this._skipUpdatingInputValue = true;
+
+            this.dispatchEventToListeners(WI.FontVariationDetailsSectionRow.Event.VariationValueChanged, {tag: this._tag, value: Math.min(Math.max(this._minimumValue, valueAsNumber), this._maximumValue)});
+            return;
+        }
+
+        this._hasValidValue = true;
+        this._skipUpdatingInputValue = false;
+        this.warningMessage = null;
+        this.value = valueAsNumber;
+
+        clearTimeout(this._showValidityTimeout);
+        this.dispatchEventToListeners(WI.FontVariationDetailsSectionRow.Event.VariationValueChanged, {tag: this._tag, value: valueAsNumber});
+    }
+
+    _handleValueKeydownEvent(event)
+    {
+        // By default, a HTML number input will accept and display any character.
+        if (event.key.length === 1 && !/[0-9\.\-]/.test(event.key)) {
+            event.preventDefault();
+            return;
+        }
+
+        let valueAsNumber = event.target.valueAsNumber;
+        let step = this._step;
+
+        if (event.shiftKey && event.altKey)
+            step = this._step;
+        else if (event.shiftKey)
+            step = this._step * 10;
+        else if (event.altKey)
+            step = this._step / 10;
+
+        // A number input will only increment and decrement within the range defined by its `min` and `max` values.
```

Comment:
this is kinda an odd way of handling it IMO

other places that we do things like this instead just do a `Number.constrain(valueAsNumber + step, this._minimumValue, this._maximumValue)`

i.e. i dont think we need to muck with the `this._valueInputElement.step` on every `"keydown"`

Reply:
There are two other places where the modifier keys change the step:

- [BoxShadowEditor.js#L303-L324](https://github.com/WebKit/WebKit/blob/main/Source/WebInspectorUI/UserInterface/Views/BoxShadowEditor.js#L303-L324)
Operating on an `<input type="text">`
- [SpreadsheetTextField.js#L285-L295](https://github.com/WebKit/WebKit/blob/main/Source/WebInspectorUI/UserInterface/Views/SpreadsheetTextField.js#L285-L295)
Operating on a `<span contenteditable>`

Neither is working on a `<input "type=number">` that natively handles increment/decrement, and which we're trying to use here.

They're not consistent either, one assumes a default step of `1`. For the variation axis, the default step can be `0.1` or `0.01` for axes with very short ranges ([the Skia font](https://v-fonts.com/fonts/skia) is a good example).

As a follow-up we're considering making a custom element that wraps the native HTML number input with the behaviors and constraints we expect in Web Inspector. It might not replace the `<span contenteditable>` case which handles a lot of extra behavior, but it can serve as a predictable number input for other uses, like this one in the Fonts sidebar panel.

Reply:
I tried it this way and it adds a host of other issues. 

The number input already handles increment/decrement on ArrowUp/ArrowDown with the `step` defined on the element.

We can't set `this._valueInputElement.value` and then `event.preventDefault()` after our own mutation to counteract that because it will prevent triggering the "input" event handler.

If we want to mutate the value, we have to account for the browser's native increment/decrement behavior: subtract or add the input's `step`. With a variable `step` to deal with that's just a lot of needless complexity.

I prefer to give the input element its desired step and leave the browser to do the work.

https://user-images.githubusercontent.com/105809798/222537992-690e735d-27ea-4999-896e-942c5116d15d.mov

Reply:
yeah that's kinda what i was trying to get at (i.e. dont even use HTML `step` and instead do the work ourselves)

Reply:
The number input won't allow a step of zero. 😮‍💨  

Frankly, this is too much fighting the native input. If it doesn't do what we need, might as well replace it.
I used a text input instead and explicitly `parseFloat()` on the value to get the number out.

In a follow-up, I'll generalize this mock-number input with the other ones we use.

Aside: someone else's [lament](https://stackoverflow.blog/2022/12/26/why-the-number-input-is-the-worst-input/) on the inadequacy of the number input. Some lessons need to be experience to be learned. Now I understand why we don't use the builtin number input elsewhere in the code.